README.txt
1- What is this?
<<<<<<< HEAD
	* This is a Joomla! installation/upgrade package to version 3.x
	* Joomla! Official site: http://www.joomla.org
	* Joomla! 3.4 version history - https://docs.joomla.org/Joomla_3.4_version_history
	* Detailed changes in the Changelog: https://github.com/joomla/joomla-cms/commits/master
=======
    * This is a Joomla! installation/upgrade package to version 2.5.x
    * Joomla! Official site: http://www.joomla.org
    * Joomla 2.5 version history - http://docs.joomla.org/Joomla_2.5_version_history
    * Detailed changes in the Changelog: https://github.com/joomla/joomla-cms/commits/2.5.x
>>>>>>> FETCH_HEAD

2- What is Joomla?
	* Joomla is a Content Management System (CMS) which enables you to build Web sites and powerful online applications.
	* It's a free and OpenSource software, distributed under the GNU General Public License version 2 or later
	* This is a simple and powerful web server application and it requires a server with PHP and either MySQL, PostgreSQL, or SQL Server to run it.
	More details here: http://www.joomla.org/about-joomla.html

3- Is Joomla for you?
	* Joomla is the right solution for any content web project: https://docs.joomla.org/Joomla_Is_it_for_me%3F
	* See Features - http://www.joomla.org/core-features.html
	* Try out our online demo: https://demo.joomla.org/

4- How to find a Joomla! translation?
	* Repository of accredited language packs: http://community.joomla.org/translations.html
	* You can also add languages directly to your website via your Joomla! administration panel.

5- Learn Joomla!
	* Read Getting Started with Joomla to find out the basics: https://docs.joomla.org/Getting_Started_with_Joomla!
	* Before installing, read the beginners guide: https://docs.joomla.org/Beginners

6- What are the limits of Joomla?
	* Joomla sites can be extended in functionalities with Extensions that you can create (or download) to suite your needs.
	* There are lots of ready made extensions that you can download and install.
	* See the Joomla! Extensions Directory (JED): http://extensions.joomla.org

7- Is it easy to change the layout display?
	* The layout is controlled by templates that you can edit.
	* There are lots of ready made templates that you can download.

8- Ready to install Joomla?
<<<<<<< HEAD
	* See minimum requirements here: http://www.joomla.org/technical-requirements.html
	* How do you install Joomla! ? - https://docs.joomla.org/Installing_Joomla!
	* Start your Joomla experience building your site with a local test server.
	When ready it can be moved to an on-line hosting account of your choice.
	See the tutorial: https://docs.joomla.org/Tutorial:Joomla_Local_install
=======
    * See minimum requirements here: http://www.joomla.org/technical-requirements.html
    * How do you install Joomla! ? - http://docs.joomla.org/J2.5:Installing_Joomla
    * Start your Joomla experience building your site with a local test server.
	When ready it can be moved to an on-line hosting account of your choice. 
	See the tutorial: http://docs.joomla.org/Tutorial:Joomla_Local_install
>>>>>>> FETCH_HEAD

9- Updates are free!
	* Always use the latest version: http://www.joomla.org/download.html

10- Where can you get support and help?
	* The Joomla! Documentation: https://docs.joomla.org/Main_Page
	* FAQ Frequently Asked Questions: https://docs.joomla.org/Category:FAQ
	* Find the information you need: https://docs.joomla.org/Start_here
	* Find help and other users: http://www.joomla.org/about-joomla/create-and-share.html
	* Post questions at our forums: http://forum.joomla.org
	* Joomla Resources Directory (JRD): http://resources.joomla.org/tos.html

11- Do you already have a Joomla site that's not built with Joomla 3.x ?
	* What's new in Joomla 3.x: http://www.joomla.org/3
	* What are the main differences from 2.5 to 3? Table of contents: https://docs.joomla.org/Differences_from_Joomla_2.5_to_Joomla_3.0
	* How to migrate from 2.5.x to 3.x? Tutorial: https://docs.joomla.org/Joomla_2.5_to_3.x_Step_by_Step_Migration
	* How to migrate from 1.5.x to 3.x? Tutorial: https://docs.joomla.org/Joomla_1.5_to_3.x_Step_by_Step_Migration
	* Convert an existing Web site to Joomla: https://docs.joomla.org/How_to_Convert_an_existing_Web_site_to_a_Joomla!_Web_site

12- Do you want to improve Joomla?
	* How do you request a feature? https://docs.joomla.org/How_do_you_request_a_feature%3F
	* How do you report a bug? https://docs.joomla.org/Filing_bugs_and_issues
	* Get Involved: Joomla! is a community developed software. Join the community at http://www.joomla.org
	* Are you a Developer? https://docs.joomla.org/Developers
	* Are you a Web designer? https://docs.joomla.org/Web_designers

Copyright:
<<<<<<< HEAD
	* Copyright (C) 2005 - 2015 Open Source Matters. All rights reserved.
	* Credits: https://docs.joomla.org/Joomla!_Credits_and_Thanks
	* Distributed under the GNU General Public License version 2 or later
	* See Licenses details at https://docs.joomla.org/Joomla_Licenses
=======
    * Copyright (C) 2005 - 2014 Open Source Matters. All rights reserved.
    * Credits: http://docs.joomla.org/Joomla!_Credits_and_Thanks
    * Distributed under the GNU General Public License version 2 or later
    * See Licenses details at http://docs.joomla.org/Joomla_Licenses
>>>>>>> FETCH_HEAD


tests/CHANGELOG
0. Metadata
-----------
<<<<<<< HEAD
Copyright:		Copyright (C) 2005 - 2015 Open Source Matters, Inc. All rights reserved.
=======
Copyright:		Copyright (C) 2005 - 2014 Open Source Matters, Inc. All rights reserved.
>>>>>>> FETCH_HEAD
License:		GNU General Public License version 2 or later; see LICENSE.txt

1. Copyright and disclaimer
---------------------------
This application is opensource software released under the GPL.  Please
see source code and the LICENSE file


2. Changelog
------------
This is a non-exhaustive (but still near complete) changelog for
Joomla! automated tests, including alpha and beta versions.
Our thanks to all those people who've contributed bug reports and
code fixes.

Legend:

* -> Security Fix
# -> Bug Fix
$ -> Language fix or change
+ -> Addition
^ -> Change
- -> Removed
! -> Note

<<<<<<< HEAD
05-Dec-2013 Mark Dexter
 # Clean up some webdriver tests to hopefully fix intermittent problems.

14-Nov-2013 Mark Dexter
 # Fix problem with button not being visible on page.
  
01-Nov-2013 Mark Dexter
 # Fix problem with hidden filters in User Manager.

31-Oct-2013 Mark Dexter
 # Fix problem with category filter in article manager.

30-Oct-2013 Mark Dexter
 # Fix weblink and newsfeed edit for markup changes.

29-Oct-2013 Mark Dexter
 # More webdriver fixes for 3.2.

27-Oct-2013 Mark Dexter
 # More webdriver fixes for 3.2 changes. Most tests working again.

18-Oct-2013 Mark Dexter
 # Fix Webdriver tests for 3.2 changes.

09-Oct-2013 Mark Dexter
 # Fix Webdriver page class for markup changes.

03-Oct-2013 Mark Dexter
 # Fix system tests for code changes.
 # Fix another system test.

14-Sep-2013 Mark Dexter
 + Add system tests for categories and articles in the front end. Thanks Puneet Kala.

11-Sep-2013 Mark Dexter
 # Fix failing system tests.

04-Sep-2013 Mark Dexter
 + Add some more front-end system tests. Thanks Puneet Kala.

01-Sep-2013 Mark Dexter
 # Fix problem with renamed test method.
 # Fix failing system test due to new option in Global Configuration.

27-Aug-2013 Mark Dexter
 + Add some system tests for front end. Thanks Puneet Kala!

21-Aug-2013 Mark Dexter
 + Start new front-end tests for articles. Thanks Puneet Kala!

16-Aug-2013 Michael Babker
 ^ Add a conditional for running app install

15-Aug-2013 Mark Dexter
 # Fix failing system tests due to markup changes.

09-Aug-2013 Mark Dexter
 # Fix failing system tests due to markup changes.

31-Jul-2013 Michael Babker
 + Add test coverage to JPlugin constructor testing private vars
 + Add test coverage to JLoader::registerPrefix()

27-Jul-2013 Michael Babker
 + Add test coverage to JViewLegacy::getName()

24-Jul-2013 Mark Dexter
 + New Webdriver tests for changing article state and front-end login. Thanks Puneet Kala.

18-Jul-2013 Mark Dexter
 # Fix problem with failing system tests.

16-Jul-2013 Mark Dexter
 + Add webdriver tests for some components and extensions. Thanks Puneet Kala.

07-Jul-2013 Mark Dexter
 # Refactor delete method in webdriver tests
 # Fix failing system tests due to markup changes.

20-Jun-2013 Mark Dexter
 # Fix Webdriver tests for Google Chrome, markup changes.

15-Jun-2013 Mark Dexter
 # Fix failing system test.

11-Jun-2013 Mark Dexter
 # Fix problems with system tests from changes in HTML markup.

02-Jun-2013 Michael Babker
 + Port Feed, Filesystem, Filter, and Form package unit tests from Platform
 + Port Github package unit tests from Platform
 + Port Google and Grid package unit tests from Platform
 + Port Image, Keychain, and Language package unit tests from Platform
 + Port Mail, Mediawiki, Model, and Object package unit tests from Platform
 + Port Plugin, Profiler, Registry, String, and Twitter package unit tests from Platform
 + Port User, Utilities, and View package unit tests from Platform
 ^ Refactor unit tests to use CSV datasets
 + Port legacy tree unit tests from Platform

01-Jun-2013 Michael Babker
 + Port Database package unit tests from Platform
 + Port URI package unit tests from Platform
 + Port Facebook package unit tests from Platform
 + Port Linkedin, Oauth1, OAuth2, and Openstreetmap package unit tests from Platform
 + Port Access and Application package unit tests from Platform
 + Port Archive, Base, Cache, Client, Controller, and Crypt package unit tests from Platform
 + Port Data, Document, Date, Environment, and Event package unit tests from Platform

28-May-2013 Mark Dexter
 + Add webdriver tests for contacts, weblinks, and banners. Thanks George Wilson.
 + Add skeleton classes for com_categories to work around inconsistent URL issue.

22-May-2013 Mark Dexter
 # Fix failing system tests due to markup changes.

16-May-2013 Mark Dexter
 + Added TagManager0002Test.php. Thanks Puneet Kala.

09-May-2013 Mark Dexter
 # Finish system tests for back end Tags Manager. Thanks Puneet Kala.

08-May-2013 Mark Dexter
 # Fix failing system test for markup change in Template Manager.

04-May-2013 Michael Babker
 # Fix failing unit tests

03-May-2013 JM Simonet
 # Corrected test for #30788

02-May-2013 Mark Dexter
 # Fix up doc blocks as example for Webdriver page classes and tests.
 # Added test for Tag Manager. Thanks Puneet Kala.

30-Apr-2013 Mark Dexter
 # Fix system tests for clearfix class in front end

25-Apr-2013 Mark Dexter
 # Fix webdriver tests for change in module edit form (using textarea for class suffix)

22-Apr-2013 Mark Dexter
 # Change error reporting to Development in auto install.

20-Apr-2013 Mark Dexter
 # Fix failing system test from change in class suffix from input to textarea.

12-Apr-2013 Michael Babker
 # Fix failing unit tests from UCM refactor

06-Apr-2013 Mark Dexter
 # Fix failing system test.

29-Mar-2013 Michael Babker
 +# Test JTableCorecontent, fix bugs found in class

29-Mar-2013 Mark Dexter
 # Fix failing system tests.
 # More system test fixes, fix CHANGELOG typo.

26-Mar-2013 Mark Dexter
 # Fix failing system tests due to login markup change, Tags menu item change

26-Mar-2013 Michael Babker
 +# Expand Schema test coverage


libraries/loader.php
<?php
/**
 * @package    Joomla.Platform
 *
<<<<<<< HEAD
 * @copyright  Copyright (C) 2005 - 2015 Open Source Matters, Inc. All rights reserved.
=======
 * @copyright  Copyright (C) 2005 - 2014 Open Source Matters, Inc. All rights reserved.
>>>>>>> FETCH_HEAD
 * @license    GNU General Public License version 2 or later; see LICENSE
 */

defined('JPATH_PLATFORM') or die;

/**
 * Static class to handle loading of libraries.
 *
 * @package  Joomla.Platform
 * @since    11.1
 */
abstract class JLoader
{
	/**
	 * Container for already imported library paths.
	 *
	 * @var    array
	 * @since  11.1
	 */
	protected static $classes = array();

	/**
	 * Container for already imported library paths.
	 *
	 * @var    array
	 * @since  11.1
	 */
	protected static $imported = array();

	/**
	 * Container for registered library class prefixes and path lookups.
	 *
	 * @var    array
	 * @since  12.1
	 */
	protected static $prefixes = array();

	/**
	 * Holds proxy classes and the class names the proxy.
	 *
	 * @var    array
	 * @since  3.2
	 */
	protected static $classAliases = array();

	/**
	 * Holds the inverse lookup for proxy classes and the class names the proxy.
	 *
	 * @var    array
	 * @since  3.4
	 */
	protected static $classAliasesInverse = array();

	/**
	 * Container for namespace => path map.
	 *
	 * @var    array
	 * @since  12.3
	 */
	protected static $namespaces = array();

	/**
	 * Method to discover classes of a given type in a given path.
	 *
	 * @param   string   $classPrefix  The class name prefix to use for discovery.
	 * @param   string   $parentPath   Full path to the parent folder for the classes to discover.
	 * @param   boolean  $force        True to overwrite the autoload path value for the class if it already exists.
	 * @param   boolean  $recurse      Recurse through all child directories as well as the parent path.
	 *
	 * @return  void
	 *
	 * @since   11.1
	 */
	public static function discover($classPrefix, $parentPath, $force = true, $recurse = false)
	{
		try
		{
			if ($recurse)
			{
				$iterator = new RecursiveIteratorIterator(
					new RecursiveDirectoryIterator($parentPath),
					RecursiveIteratorIterator::SELF_FIRST
				);
			}
			else
			{
				$iterator = new DirectoryIterator($parentPath);
			}

			/* @type  $file  DirectoryIterator */
			foreach ($iterator as $file)
			{
				$fileName = $file->getFilename();

				// Only load for php files.
				if ($file->isFile() && $file->getExtension() == 'php')
				{
					// Get the class name and full path for each file.
					$class = strtolower($classPrefix . preg_replace('#\.php$#', '', $fileName));

					// Register the class with the autoloader if not already registered or the force flag is set.
					if (empty(self::$classes[$class]) || $force)
					{
						self::register($class, $file->getPath() . '/' . $fileName);
					}
				}
			}
		}
		catch (UnexpectedValueException $e)
		{
			// Exception will be thrown if the path is not a directory. Ignore it.
		}
	}

	/**
	 * Method to get the list of registered classes and their respective file paths for the autoloader.
	 *
	 * @return  array  The array of class => path values for the autoloader.
	 *
	 * @since   11.1
	 */
	public static function getClassList()
	{
		return self::$classes;
	}

	/**
	 * Method to get the list of registered namespaces.
	 *
	 * @return  array  The array of namespace => path values for the autoloader.
	 *
	 * @since   12.3
	 */
	public static function getNamespaces()
	{
		return self::$namespaces;
	}

	/**
	 * Loads a class from specified directories.
	 *
	 * @param   string  $key   The class name to look for (dot notation).
	 * @param   string  $base  Search this directory for the class.
	 *
	 * @return  boolean  True on success.
	 *
	 * @since   11.1
	 */
	public static function import($key, $base = null)
	{
		// Only import the library if not already attempted.
		if (!isset(self::$imported[$key]))
		{
			// Setup some variables.
			$success = false;
			$parts = explode('.', $key);
			$class = array_pop($parts);
			$base = (!empty($base)) ? $base : __DIR__;
			$path = str_replace('.', DIRECTORY_SEPARATOR, $key);

			// Handle special case for helper classes.
			if ($class == 'helper')
			{
				$class = ucfirst(array_pop($parts)) . ucfirst($class);
			}
			// Standard class.
			else
			{
				$class = ucfirst($class);
			}

			// If we are importing a library from the Joomla namespace set the class to autoload.
			if (strpos($path, 'joomla') === 0)
			{
				// Since we are in the Joomla namespace prepend the classname with J.
				$class = 'J' . $class;

				// Only register the class for autoloading if the file exists.
				if (is_file($base . '/' . $path . '.php'))
				{
					self::$classes[strtolower($class)] = $base . '/' . $path . '.php';
					$success = true;
				}
			}
			/*
			 * If we are not importing a library from the Joomla namespace directly include the
			 * file since we cannot assert the file/folder naming conventions.
			 */
			else
			{
				// If the file exists attempt to include it.
				if (is_file($base . '/' . $path . '.php'))
				{
					$success = (bool) include_once $base . '/' . $path . '.php';
				}
			}

			// Add the import key to the memory cache container.
			self::$imported[$key] = $success;
		}

		return self::$imported[$key];
	}

	/**
	 * Load the file for a class.
	 *
	 * @param   string  $class  The class to be loaded.
	 *
	 * @return  boolean  True on success
	 *
	 * @since   11.1
	 */
	public static function load($class)
	{
		// Sanitize class name.
		$class = strtolower($class);

		// If the class already exists do nothing.
		if (class_exists($class, false))
		{
			return true;
		}

		// If the class is registered include the file.
		if (isset(self::$classes[$class]))
		{
			include_once self::$classes[$class];

			return true;
		}

		return false;
	}

	/**
	 * Directly register a class to the autoload list.
	 *
	 * @param   string   $class  The class name to register.
	 * @param   string   $path   Full path to the file that holds the class to register.
	 * @param   boolean  $force  True to overwrite the autoload path value for the class if it already exists.
	 *
	 * @return  void
	 *
	 * @since   11.1
	 */
	public static function register($class, $path, $force = true)
	{
		// Sanitize class name.
		$class = strtolower($class);

		// Only attempt to register the class if the name and file exist.
		if (!empty($class) && is_file($path))
		{
			// Register the class with the autoloader if not already registered or the force flag is set.
			if (empty(self::$classes[$class]) || $force)
			{
				self::$classes[$class] = $path;
			}
		}
	}

	/**
	 * Register a class prefix with lookup path.  This will allow developers to register library
	 * packages with different class prefixes to the system autoloader.  More than one lookup path
	 * may be registered for the same class prefix, but if this method is called with the reset flag
	 * set to true then any registered lookups for the given prefix will be overwritten with the current
	 * lookup path. When loaded, prefix paths are searched in a "last in, first out" order.
	 *
	 * @param   string   $prefix   The class prefix to register.
	 * @param   string   $path     Absolute file path to the library root where classes with the given prefix can be found.
	 * @param   boolean  $reset    True to reset the prefix with only the given lookup path.
	 * @param   boolean  $prepend  If true, push the path to the beginning of the prefix lookup paths array.
	 *
	 * @return  void
	 *
	 * @throws  RuntimeException
	 *
	 * @since   12.1
	 */
	public static function registerPrefix($prefix, $path, $reset = false, $prepend = false)
	{
		// Verify the library path exists.
		if (!file_exists($path))
		{
			$path = (str_replace(JPATH_ROOT, '', $path) == $path) ? basename($path) : str_replace(JPATH_ROOT, '', $path);

			throw new RuntimeException('Library path ' . $path . ' cannot be found.', 500);
		}

		// If the prefix is not yet registered or we have an explicit reset flag then set set the path.
		if (!isset(self::$prefixes[$prefix]) || $reset)
		{
			self::$prefixes[$prefix] = array($path);
		}
		// Otherwise we want to simply add the path to the prefix.
		else
		{
			if ($prepend)
			{
				array_unshift(self::$prefixes[$prefix], $path);
			}
			else
			{
				self::$prefixes[$prefix][] = $path;
			}
		}
	}

	/**
	 * Offers the ability for "just in time" usage of `class_alias()`.
	 * You cannot overwrite an existing alias.
	 *
	 * @param   string  $alias     The alias name to register.
	 * @param   string  $original  The original class to alias.
	 *
	 * @return  boolean  True if registration was successful. False if the alias already exists.
	 *
	 * @since   3.2
	 */
	public static function registerAlias($alias, $original)
	{
		if (!isset(self::$classAliases[$alias]))
		{
			self::$classAliases[$alias] = $original;

			// Remove the root backslash if present.
			if ($original[0] == '\\')
			{
				$original = substr($original, 1);
			}

			if (!isset(self::$classAliasesInverse[$original]))
			{
				self::$classAliasesInverse[$original] = array($alias);
			}
			else
			{
				self::$classAliasesInverse[$original][] = $alias;
			}

			return true;
		}

		return false;
	}

	/**
	 * Register a namespace to the autoloader. When loaded, namespace paths are searched in a "last in, first out" order.
	 *
	 * @param   string   $namespace  A case sensitive Namespace to register.
	 * @param   string   $path       A case sensitive absolute file path to the library root where classes of the given namespace can be found.
	 * @param   boolean  $reset      True to reset the namespace with only the given lookup path.
	 * @param   boolean  $prepend    If true, push the path to the beginning of the namespace lookup paths array.
	 *
	 * @return  void
	 *
	 * @throws  RuntimeException
	 *
	 * @since   12.3
	 */
	public static function registerNamespace($namespace, $path, $reset = false, $prepend = false)
	{
		// Verify the library path exists.
		if (!file_exists($path))
		{
			$path = (str_replace(JPATH_ROOT, '', $path) == $path) ? basename($path) : str_replace(JPATH_ROOT, '', $path);

			throw new RuntimeException('Library path ' . $path . ' cannot be found.', 500);
		}

		// If the namespace is not yet registered or we have an explicit reset flag then set the path.
		if (!isset(self::$namespaces[$namespace]) || $reset)
		{
			self::$namespaces[$namespace] = array($path);
		}

		// Otherwise we want to simply add the path to the namespace.
		else
		{
			if ($prepend)
			{
				array_unshift(self::$namespaces[$namespace], $path);
			}
			else
			{
				self::$namespaces[$namespace][] = $path;
			}
		}
	}

	/**
	 * Method to setup the autoloaders for the Joomla Platform.
	 * Since the SPL autoloaders are called in a queue we will add our explicit
	 * class-registration based loader first, then fall back on the autoloader based on conventions.
	 * This will allow people to register a class in a specific location and override platform libraries
	 * as was previously possible.
	 *
	 * @param   boolean  $enablePsr       True to enable autoloading based on PSR-0.
	 * @param   boolean  $enablePrefixes  True to enable prefix based class loading (needed to auto load the Joomla core).
	 * @param   boolean  $enableClasses   True to enable class map based class loading (needed to auto load the Joomla core).
	 *
	 * @return  void
	 *
	 * @since   12.3
	 */
	public static function setup($enablePsr = true, $enablePrefixes = true, $enableClasses = true)
	{
		if ($enableClasses)
		{
			// Register the class map based autoloader.
			spl_autoload_register(array('JLoader', 'load'));
		}

		if ($enablePrefixes)
		{
			// Register the J prefix and base path for Joomla platform libraries.
			self::registerPrefix('J', JPATH_PLATFORM . '/joomla');

			// Register the prefix autoloader.
			spl_autoload_register(array('JLoader', '_autoload'));
		}

		if ($enablePsr)
		{
			// Register the PSR-0 based autoloader.
			spl_autoload_register(array('JLoader', 'loadByPsr0'));
			spl_autoload_register(array('JLoader', 'loadByAlias'));
		}
	}

	/**
	 * Method to autoload classes that are namespaced to the PSR-0 standard.
	 *
	 * @param   string  $class  The fully qualified class name to autoload.
	 *
	 * @return  boolean  True on success, false otherwise.
	 *
	 * @since   13.1
	 */
	public static function loadByPsr0($class)
	{
		// Remove the root backslash if present.
		if ($class[0] == '\\')
		{
			$class = substr($class, 1);
		}

		// Find the location of the last NS separator.
		$pos = strrpos($class, '\\');

		// If one is found, we're dealing with a NS'd class.
		if ($pos !== false)
		{
			$classPath = str_replace('\\', DIRECTORY_SEPARATOR, substr($class, 0, $pos)) . DIRECTORY_SEPARATOR;
			$className = substr($class, $pos + 1);
		}
		// If not, no need to parse path.
		else
		{
			$classPath = null;
			$className = $class;
		}

		$classPath .= str_replace('_', DIRECTORY_SEPARATOR, $className) . '.php';

		// Loop through registered namespaces until we find a match.
		foreach (self::$namespaces as $ns => $paths)
		{
			if (strpos($class, $ns) === 0)
			{
				// Loop through paths registered to this namespace until we find a match.
				foreach ($paths as $path)
				{
					$classFilePath = $path . DIRECTORY_SEPARATOR . $classPath;

					// We check for class_exists to handle case-sensitive file systems
					if (file_exists($classFilePath) && !class_exists($class, false))
					{
						return (bool) include_once $classFilePath;
					}
				}
			}
		}

		return false;
	}

	/**
	 * Method to autoload classes that have been aliased using the registerAlias method.
	 *
	 * @param   string  $class  The fully qualified class name to autoload.
	 *
	 * @return  boolean  True on success, false otherwise.
	 *
	 * @since   3.2
	 */
	public static function loadByAlias($class)
	{
		// Remove the root backslash if present.
		if ($class[0] == '\\')
		{
			$class = substr($class, 1);
		}

		if (isset(self::$classAliases[$class]))
		{
			// Force auto-load of the regular class
			class_exists(self::$classAliases[$class], true);

			// Normally this shouldn't execute as the autoloader will execute applyAliasFor when the regular class is
			// auto-loaded above.
			if (!class_exists($class, false) && !interface_exists($class, false))
			{
				class_alias(self::$classAliases[$class], $class);
			}
		}
	}

	/**
	 * Applies a class alias for an already loaded class, if a class alias was created for it.
	 *
	 * @param   string  $class  We'll look for and register aliases for this (real) class name
	 *
	 * @return  void
	 *
	 * @since   3.4
	 */
	public static function applyAliasFor($class)
	{
		// Remove the root backslash if present.
		if ($class[0] == '\\')
		{
			$class = substr($class, 1);
		}

		if (isset(self::$classAliasesInverse[$class]))
		{
			foreach (self::$classAliasesInverse[$class] as $alias)
			{
				class_alias($class, $alias);
			}
		}
	}

	/**
	 * Autoload a class based on name.
	 *
	 * @param   string  $class  The class to be loaded.
	 *
	 * @return  boolean  True if the class was loaded, false otherwise.
	 *
	 * @since   11.3
	 */
	private static function _autoload($class)
	{
		foreach (self::$prefixes as $prefix => $lookup)
		{
<<<<<<< HEAD
			$chr = strlen($prefix) < strlen($class) ? $class[strlen($prefix)] : 0;

			if (strpos($class, $prefix) === 0 && ($chr === strtoupper($chr)))
=======
			if (strpos($class, $prefix) === 0 && self::isFirstLetterAfterPrefixUpperCase($prefix, $class))
>>>>>>> FETCH_HEAD
			{
				return self::_load(substr($class, strlen($prefix)), $lookup);
			}
		}

		return false;
	}

	/**
	 * Check if first letter after the prefix length is uppercase (First letter of class name)
	 * This allows to match proper library when one library prefix is included in the prefix of another class
	 * e.g: LMyclass and LibMyclass
	 *
	 * @param   string  $prefix  Library prefix
	 * @param   string  $class   Classname
	 *
	 * @return bool
	 */
	private static function isFirstLetterAfterPrefixUpperCase($prefix, $class)
	{
		if (strlen($prefix) >= strlen($class))
		{
			return false;
		}

		$chr = $class[strlen($prefix)];

		return $chr === strtoupper($chr);
	}

	/**
	 * Load a class based on name and lookup array.
	 *
	 * @param   string  $class   The class to be loaded (wihtout prefix).
	 * @param   array   $lookup  The array of base paths to use for finding the class file.
	 *
	 * @return  boolean  True if the class was loaded, false otherwise.
	 *
	 * @since   12.1
	 */
	private static function _load($class, $lookup)
	{
		// Split the class name into parts separated by camelCase.
		$parts = preg_split('/(?<=[a-z0-9])(?=[A-Z])/x', $class);

		// If there is only one part we want to duplicate that part for generating the path.
		$parts = (count($parts) === 1) ? array($parts[0], $parts[0]) : $parts;

		foreach ($lookup as $base)
		{
			// Generate the path based on the class name parts.
			$path = $base . '/' . implode('/', array_map('strtolower', $parts)) . '.php';

			// Load the file if it exists.
			if (file_exists($path))
			{
				return include $path;
			}
		}

		return false;
	}
}

/**
 * Global application exit.
 *
 * This function provides a single exit point for the platform.
 *
 * @param   mixed  $message  Exit code or string. Defaults to zero.
 *
 * @return  void
 *
 * @codeCoverageIgnore
 * @since   11.1
 */
function jexit($message = 0)
{
	exit($message);
}

/**
 * Intelligent file importer.
 *
 * @param   string  $path  A dot syntax path.
 *
 * @return  boolean  True on success.
 *
 * @since   11.1
 */
function jimport($path)
{
	return JLoader::import($path);
}
